File: macos/QMK Toolbox/USB/Bootloader/CaterinaDevice.m
#import "CaterinaDevice.h"

@implementation CaterinaDevice

- (id)initWithUSBDevice:(USBDevice *)usbDevice {
    if (self = [super initWithUSBDevice:usbDevice]) {
        self.name = @"Caterina";
        self.type = BootloaderTypeCaterina;
        self.eepromFlashable = YES;
        while (self.serialPort == nil) {
            self.serialPort = [self.usbDevice findSerialPort];
        }
    }
    return self;
}

-(void)flashWithMCU:(NSString *)mcu file:(NSString *)file {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"avr109", @"-U", [NSString stringWithFormat:@"flash:w:%@:i", file], @"-P", self.serialPort, @"-C", @"avrdude.conf"]];
}

-(void)flashEEPROMWithMCU:(NSString *)mcu file:(NSString *)file {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"avr109", @"-U", [NSString stringWithFormat:@"eeprom:w:%@:i", file], @"-P", self.serialPort, @"-C", @"avrdude.conf"]];
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ [%@]", [super description], self.serialPort];
}

@end

File: macos/QMK Toolbox/USB/Bootloader/CaterinaDevice.swift
import Foundation

class CaterinaDevice: BootloaderDevice {
    private var serialPort: String?

    override init(usbDevice: USBDevice) {
        super.init(usbDevice: usbDevice)
        name = "Caterina"
        type = .caterina
        eepromFlashable = true
        while (serialPort == nil) {
            serialPort = findSerialPort()
        }
    }

    override func flash(_ mcu: String, file: String) {
        runProcess("avrdude", args: ["-p", mcu, "-c", "avr109", "-U", "flash:w:\(file):i", "-P", serialPort!, "-C", "avrdude.conf"])
    }

    override func flashEEPROM(_ mcu: String, file: String) {
        runProcess("avrdude", args: ["-p", mcu, "-c", "avr109", "-U", "eeprom:w:\(file):i", "-P", serialPort!, "-C", "avrdude.conf"])
    }

    override var description: String {
        "\(super.description) [\(serialPort!)]"
    }
}
